### PR TITLE
Update NUnit3TestAdapter version to 4.0.0-beta.2

### LINKED ISSUE #20509: [BUG] Test runner process crashes due to OutOfMemoryException
**Describe the bug**
Test runs in the 'net - search - tests' pipeline are failing frequently due to the test host process crashing with OutOfMemoryException when parsing result files.

Some example runs -

- https://dev.azure.com/azure-sdk/internal/_build/results?buildId=847015&view=results
- https://dev.azure.com/azure-sdk/internal/_build/results?buildId=844514&view=results
- https://dev.azure.com/azure-sdk/internal/_build/results?buildId=842006&view=results

**Expected behavior**
The results file is generated successfully after test execution.

**Actual behavior (include Exception or Stack Trace)**
Test runner process crashes when processing results.

**To Reproduce**
See the above build runs for details.

### PR DESCRIPTION
**Pull request by Mohit-Chakraborty**

### CODE REVIEW

**Comment by Mohit-Chakraborty** [comment]
@mikeharder - I ran the live search tests and they passed. I think the change to update the NUnit test adapter has helped.
What other tests should I run here?

**Comment by mikeharder** [comment]
@Mohit-Chakraborty: I triggered runs of core and storage tests, if these pass I think this is safe enough to merge.  If we find any failures after merge we can investigate.

**Comment by pakrym** [comment]
Any particular reason to get the update especially a beta?

**Comment by mikeharder** [comment]
@pakrym: The latest RTM version is crashing with OOM per this issue: https://github.com/Azure/azure-sdk-for-net/issues/20509.  I don't think @Mohit-Chakraborty has been able to repro locally, so we're not sure exactly what's causing it.  But I suspect it's an issue in NUnit3TestAdapter, so I suggested we try the latest version and it does seem to fix the OOM.

If you would rather not use a Beta, we could test GA versions older than 3.17.0 to see if any of them fix the OOM as well.

**Comment by pakrym** [comment]
As far as I know, 4.0.0-* is a major rewrite of NUnit. I'm not strictly against it but think we should be careful taking it.

Doesn't `Test host process crashed` apply to the child `testhost` process?

**Comment by pakrym** [comment]
Yep, use this test:

```C#
using System;
using System.Threading;
using NUnit.Framework;
using System.Collections.Generic;

namespace ooomunit
{
    public class Tests
    {
        [Test]
        public void Test1()
        {
            ThreadPool.QueueUserWorkItem(_ => 
            {
            List<byte[]> b = new();
            while (true)
            {
                b.Add(new byte[100000000]);
            }
            }
            );
            
            Thread.Sleep(1000000);
        }
    }
}
```

**Comment by pakrym** [comment]
I would also look into the `--blame-*` set of parameters for dotnet test. https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-test

File: eng/Packages.Data.props [hunk 1]
```diff
@@ -208,7 +208,7 @@
     <PackageReference Update="Newtonsoft.Json" Version="10.0.3" />
     <PackageReference Update="NSubstitute" Version="3.1.0" />
     <PackageReference Update="NUnit" Version="3.13.2" />
-    <PackageReference Update="NUnit3TestAdapter" Version="3.17.0" />
+    <PackageReference Update="NUnit3TestAdapter" Version="4.0.0-beta.2" />
```
Comment (weshaggard) [inline eng/Packages.Data.props:211]: Similar to @pakrym's feedback I would like to avoid taking a dependency on a prerelease if it can be helped.

**Comment by mikeharder** [comment]
I agree we need to step back and understand what is happening here.  If this is indeed a test or product bug causing OOM, then why **doesn't** it repro with the newer `NUnit3TestAdapter`?

**Comment by pakrym** [comment]
>. If this is indeed a test or product bug causing OOM, then why doesn't it repro with the newer NUnit3TestAdapter?

Maybe some changes in test ordering/parallelization? I think adding an ability (or learning a way) to collect test crash dumps is a great exercise in general.

**Pull request closed without merging**